taskcluster/translations_taskgraph/transforms/upload_artifacts.py

```diff
@@ -0,0 +1,153 @@
+from taskgraph.util.schema import resolve_keyed_by
+
+from translations_taskgraph.util.dependencies import get_upload_artifacts_upstream_dependency
+from translations_taskgraph.util.substitution import substitute
+
+transforms = TransformSequence()
+
+
+@transforms.add
+def evaluate_keyed_by(_, jobs):
+    for job in jobs:
+        upstream_kind = get_upload_artifacts_upstream_dependency(job["dependencies"])[0]
+        for field in ("upstream-artifacts", "artifact-map"):
+            resolve_keyed_by(
+                job["worker"],
+                field,
+                item_name=job["description"],
+                **{
+                    "upstream-kind": upstream_kind,
+                },
+            )
+
+        yield job
+
+
+@transforms.add
+def substitute_step_dir(_, jobs):
+    for job in jobs:
+        upstream_task = get_upload_artifacts_upstream_dependency(job["dependencies"])[1]
+        step_dir = task_to_step_dir(upstream_task)
+        substitute(job["worker"]["artifact-map"], step_dir=step_dir)
+        yield job
+
+
+# tasks for where the full task label is needed to avoid overwriting logs
+# from different, eg: datasets
+FULL_TASK_LABEL_STEP_DIR_TASKS = {
+    "corpus-analyze-mono",
+    "corpus-analyze-parallel",
+    "corpus-clean-mono",
+    "corpus-clean-parallel-bicleaner-ai",
+    "corpus-clean-parallel",
+}
+TASKS_WITH_MATCHING_STEP_DIR_PREFIXES = {
+    "backtranslations-mono-trg-chunk",
+    "backtranslations-mono-trg-dechunk-translations",
+    "corpus-align-backtranslations",
+    "corpus-align-distillation",
+    "corpus-align-parallel",
+    "corpus-clean-parallel-fetch-bicleaner-model",
+    "corpus-merge-devset",
+    "corpus-merge-distillation",
+    "corpus-merge-mono-src",
+    "corpus-merge-mono-trg",
+    "corpus-merge-parallel",
+    "distillation-corpus-build-shortlist",
+    "distillation-corpus-final-filtering",
+    "distillation-mono-src-chunk",
+    "distillation-mono-src-dechunk-translations",
+    "distillation-parallel-src-chunk",
+    "distillation-parallel-src-dechunk-translations",
+    "distillation-parallel-src-translations-score",
+}
+
+
+def task_to_step_dir(task_label: str) -> str:
+    # Process these first, because we want to prioritize `corpus-clean-parallel-fetch-bicleaner-model`
+    # not having SRC-TRG in its step directory.
+    if prefixes := [
+        prefix for prefix in TASKS_WITH_MATCHING_STEP_DIR_PREFIXES if task_label.startswith(prefix)
+    ]:
+        # there should only be only be one match!
+        assert len(prefixes) == 1
+        return prefixes[0]
+    elif any([task_label.startswith(prefix) for prefix in FULL_TASK_LABEL_STEP_DIR_TASKS]):
+        return task_label
+    elif task_label.startswith("backtranslations-mono-trg-translate"):
+        n = get_chunk_number(task_label)
+        return f"backtranslations-mono-trg-translate{n}"
+    elif task_label.startswith("backtranslations-train-backwards-model"):
```

Comment:
nit: Having 2 map dicts up top might make this a bit more readable
Something like:
```
CHUNKED_TASK_LABELS = {"backtranslations-mono-trg-translate": "backtranslations-mono-trg-translate{n}"...}
TASK_LABELS = {"build-vocab": "vocab"...}
...
```
Then you could just lookup the 2 maps without having all these elifs